### DISCUSSION TITLE
How to use knex in migrations?

### DISCUSSION THREAD

**Discussion by karimdaghari** [mikro-orm/mikro-orm#2773] [Q&A] [answered]
Hi Martin,

First off, thanks a lot of the **AWESOME** work you're doing (it's an understatement tbh!).

Question is in the title. I've been looking through the docs, couldn't really find anything (besides an unpractical [reference](https://mikro-orm.io/docs/migrations#:~:text=The%20Migration.addSql()%20method%20also%20accepts%20instances%20of%20knex.%20Knex%20instance%20can%20be%20accessed%20via%20Migration.getKnex())).

Here's what I've tried: it seemed the most sound to me yet didn't work.

```ts
import { Migration } from "@mikro-orm/migrations";

export class Migration20220216182214 extends Migration {
  knex = this.driver.getConnection().getKnex();

  async up(): Promise<void> {
    this.knex.raw(
      "CREATE TABLE 'bla' (LIKE 'bloo' INCLUDING INDEX INCLUDING DEFAULTS INCLUDING)"
    );

    this.knex.schema.alterTable("bob", t => {
      t.dropForeign("x_id");
      t.foreign("x_id").references("bla.id").onDelete("cascade");
    });
  }

  async down(): Promise<void> {
    this.knex.schema.alterTable("bob", t => {
      t.dropForeign("x_id");
      t.foreign("x_id").references("bloo.id").onDelete("cascade");
    });

    this.knex.schema.dropTableIfExists("bla");
  }
}

```

_P.S: I'm aware I can use `this.addSql()` but I'd rather use a programmatic API._

**Comment by B4nan** [comment]
You have knex instance accessible via `this.getKnex()`, your example looks good to me, you just need to await those calls.

**Reply by karimdaghari** [reply]
I did. and I got a knex timemout error. 

`KnexTimeoutError: Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?`

I looked around, fiddled with the pool config but it just wouldn't work.

**Reply by B4nan** [reply]
You need to add something like `knex.transacting(this.ctx)` probably. You will have the TX context available there so you can reuse it.

**Reply by B4nan** [reply]
You could actually just work with `this.ctx`, it is a knex instance that bares the TX context. 

I guess we should return `this.ctx ?? this.driver.getConnection('write').getKnex();` from the `getKnex()` method.

**Reply by karimdaghari** [reply]
okay, give me a minute to try it out (btw, I really appreciate you taking the time to help out!)

**Comment by karimdaghari** [comment, marked as answer]
Thanks a lot @B4nan! Got it to work thanks to you.

I've included the fixed code below in case someone encounters the same issue: basically the call for the knex client has to be setup from _within_ the up/down methods and not as a class prop. 

As for the reason why, I suppose it's because the context/driver is set through the constructor/when the class in instantiated and so when we define a knex prop it's calling on a driver/ctx that hasn't been defined yet. 

_(This is a theory I just made up on the spot after taking a quick glance at the `Migration` class. Martin: please correct me if I'm wrong as I'm genuinely curious and want to understand why it didn't work.)_

```ts
import { Migration } from "@mikro-orm/migrations";

export class Migration20220216182214 extends Migration {
  async up(): Promise<void> {
    const knex = this.ctx ?? this.driver.getConnection("write").getKnex();

    knex.raw(
      "CREATE TABLE 'bla' (LIKE 'bloo' INCLUDING INDEX INCLUDING DEFAULTS)"
    );

    knex.schema.alterTable("bob", t => {
      t.dropForeign("x_id");
      t.foreign("x_id").references("bla.id").onDelete("cascade");
    });
  }

  async down(): Promise<void> {
    const knex = this.ctx ?? this.driver.getConnection("write").getKnex();

    knex.schema.alterTable("bob", t => {
      t.dropForeign("x_id");
      t.foreign("x_id").references("bloo.id").onDelete("cascade");
    });

    knex.schema.dropTableIfExists("bla");
  }
}

```
